sdk/messaging/azservicebus/internal/exported/error.go

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package exported
+
+import "fmt"
+
+// Code is an error code, usable by consuming code to work with
+// programatically.
+type Code string
+
+const (
+	// CodeConnectionLost means our connection was lost and all retry attempts failed.
+	// This typicall reflects an extended outage or connection disruption and may
+	// require manual intervention.
+	CodeConnectionLost Code = "connlost"
+
+	// CodeLockLost means that the lock token you have for a message has expired.
+	// This message will be available again after the lock period expires, or, potentially
+	// go to the dead letter queue if delivery attempts have been exceeded.
+	CodeLockLost Code = "locklost"
+)
+
+// Error represents a Service Bus specific error.
+type Error struct {
+	// Code is a stable error code which can be used as part of programatic error handling.
+	// The codes can expand in the future, but the values (and their meaning) will remain the same.
+	Code     Code
+	innerErr error
+}
+
+// Error is an error message containing the code and a user friendly message, if any.
+func (e Error) Error() string {
+	return fmt.Sprintf("(%s): %s", e.Code, e.innerErr.Error())
+}
+
+// Unwrap is implemented so this error can be used with errors.As()
+func (e Error) Unwrap() error {
```

Comment:
When a type doesn't define `Unwrap`, `errors.As` just doesn't call it and the error chain ends with that type. So, you need this method only if you want to enable applications to access `e.innerErr` via `errors.As`. I guess you don't, because `e.innerErr` will be an unexported type anyway?

Reply:
This is intentional as we're only just providing a basic wrapper. The real details are always in the underlying error, but the only part we're on the hook for, programmatically, is providing the code.

Reply:
Implementing `Unwrap()` means that the wrapped error implicitly becomes part of the error contract, as `errors.As()` and friends will have access to it.  Is this really necessary?  We've not been implementing `Unwrap()` in other error types.

Reply:
Unwrap() has now been removed.